imu_integrator/plugin.cpp

```diff
@@ -0,0 +1,479 @@
+#include <chrono>
+#include <iomanip>
+#include <thread>
+#include <eigen3/Eigen/Dense>
+
+#include "../open_vins/ov_msckf/src/state/Propagator.h"
+#include "../open_vins/ov_core/src/types/IMU.h"
+
+#include "common/plugin.hpp"
+#include "common/switchboard.hpp"
+#include "common/data_format.hpp"
+#include "common/threadloop.hpp"
+
+using namespace ILLIXR;
+
+class imu_integrator : public threadloop {
+public:
+	imu_integrator(std::string name_, phonebook* pb_)
+		: threadloop{name_, pb_}
+		, sb{pb->lookup_impl<switchboard>()}
+		, _m_imu_cam{sb->subscribe_latest<imu_cam_type>("imu_cam")}
+		, _m_in{sb->subscribe_latest<imu_integrator_seq>("imu_integrator_seq")}
+		, _m_imu_integrator_input{sb->subscribe_latest<imu_integrator_input>("imu_integrator_input")}
+		, _m_imu_raw{sb->publish<imu_raw_type>("imu_raw")}
+		, _seq_expect(1)
+	{}
+
+	virtual skip_option _p_should_skip() override {
+		auto in = _m_in->get_latest_ro();
+		if (!in || in->seq == _seq_expect-1) {
+			// No new data, sleep to keep CPU utilization low
+			std::this_thread::sleep_for(std::chrono::milliseconds{1});
+			return skip_option::skip_and_yield;
+		} else {
+			if (in->seq != _seq_expect) {
+				_stat_missed = in->seq - _seq_expect;
+			} else {
+				_stat_missed = 0;
+			}
+			_stat_processed++;
+			_seq_expect = in->seq+1;
+			return skip_option::run;
+		}
+	}
+
+	void _p_one_iteration() override {
+		const imu_cam_type *datum = _m_imu_cam->get_latest_ro();
+		double timestamp_in_seconds = (double(datum->dataset_time) / NANO_SEC);
+
+		ov_msckf::Propagator::IMUDATA data;
+        data.timestamp = timestamp_in_seconds;
+        data.wm = (datum->angular_v).cast<double>();
+        data.am = (datum->linear_a).cast<double>();
+		_imu_vec.emplace_back(data);
+
+		clean_imu_vec(timestamp_in_seconds);
+        propagate_imu_values(timestamp_in_seconds, datum->time);
+	}
+
+private:
+	const std::shared_ptr<switchboard> sb;
+
+	// IMU Data, Sequence Flag, and State Vars Needed
+	std::unique_ptr<reader_latest<imu_cam_type>> _m_imu_cam;
+	std::unique_ptr<reader_latest<imu_integrator_seq>> _m_in;
+	std::unique_ptr<reader_latest<imu_integrator_input>> _m_imu_integrator_input;
+
+	// IMU Biases
+	std::unique_ptr<writer<imu_raw_type>> _m_imu_raw;
+	std::vector<ov_msckf::Propagator::IMUDATA> _imu_vec;
+	double last_imu_offset;
+	bool has_last_offset = false;
+	long long _seq_expect, _stat_processed, _stat_missed;
+
+	int counter = 0;
+	int cam_count = 0;
+	int total_imu = 0;
+	double last_cam_time = 0;
+
+	// Open_VINS cleans IMU values older than 20 seconds, we clean values older than 5 seconds
+	void clean_imu_vec(double timestamp) {
+		auto it0 = _imu_vec.begin();
+        while (it0 != _imu_vec.end()) {
+            if (timestamp-(*it0).timestamp > 5) {
```

Comment:
Probably best that this can be configurable at some point. Or, at least use a `IMU_SAMPLE_LIFETIME` macro or something similar.